arduino/sketch/sketch_test.go

```diff
@@ -17,158 +17,375 @@ package sketch
+
+	sketchPath = paths.New("testdata", "SketchSimple", "SketchSimple.ino")
+	files = CheckForPdeFiles(sketchPath)
+	require.Empty(t, files)
+
+	sketchPath = paths.New("testdata", "SketchPde", "SketchPde.pde")
+	files = CheckForPdeFiles(sketchPath)
+	require.Len(t, files, 1)
+	require.Equal(t, sketchPath.Parent().Join("SketchPde.pde"), files[0])
+
+	sketchPath = paths.New("testdata", "SketchMultipleMainFiles", "SketchMultipleMainFiles.ino")
+	files = CheckForPdeFiles(sketchPath)
+	require.Len(t, files, 1)
+	require.Equal(t, sketchPath.Parent().Join("SketchMultipleMainFiles.pde"), files[0])
+
+	sketchPath = paths.New("testdata", "SketchMultipleMainFiles", "SketchMultipleMainFiles.pde")
+	files = CheckForPdeFiles(sketchPath)
+	require.Len(t, files, 1)
+	require.Equal(t, sketchPath.Parent().Join("SketchMultipleMainFiles.pde"), files[0])
+}
+
+func TestNewSketchWithSymlink(t *testing.T) {
+	sketchPath, _ := paths.New("testdata", "SketchWithSymlink").Abs()
+	mainFilePath := sketchPath.Join("SketchWithSymlink.ino")
+	helperFilePath := sketchPath.Join("some_folder", "helper.h")
+	helperFileSymlinkPath := sketchPath.Join("src", "helper.h")
+	srcPath := sketchPath.Join("src")
+
+	// Create a symlink in the Sketch folder
+	os.Symlink(sketchPath.Join("some_folder").String(), srcPath.String())
+	defer srcPath.Remove()
+
+	sketch, err := New(sketchPath)
+	require.NoError(t, err)
+	require.NotNil(t, sketch)
+	require.True(t, sketch.MainFile.EquivalentTo(mainFilePath))
+	require.True(t, sketch.FullPath.EquivalentTo(sketchPath))
+	require.Equal(t, sketch.OtherSketchFiles.Len(), 0)
+	require.Equal(t, sketch.AdditionalFiles.Len(), 2)
+	require.True(t, sketch.AdditionalFiles.Contains(helperFilePath))
+	require.True(t, sketch.AdditionalFiles.Contains(helperFileSymlinkPath))
+	require.Equal(t, sketch.RootFolderFiles.Len(), 0)
+}
+
+func TestNewSketchWithSymlinkLoop(t *testing.T) {
+	sketchPath, _ := paths.New("testdata", "SketchWithSymlinkLoop").Abs()
+	someSymlinkPath := sketchPath.Join("some_folder", "some_symlink")
+
+	// Create a recursive Sketch symlink
+	os.Symlink(sketchPath.String(), someSymlinkPath.String())
+	defer someSymlinkPath.Remove()
+
+	// The failure condition is New() never returning, testing for which requires setting up a timeout.
+	done := make(chan bool)
+	var sketch *Sketch
+	var err error
+	go func() {
+		sketch, err = New(sketchPath)
+		done <- true
+	}()
+
+	assert.Eventually(
+		t,
+		func() bool {
+			select {
+			case <-done:
+				return true
+			default:
+				return false
+			}
+		},
+		20*time.Second,
+		10*time.Millisecond,
+		"Infinite symlink loop while loading sketch",
+	)
+	require.Error(t, err)
+	require.Nil(t, sketch)
+}
+
+func TestSketchWithMultipleSymlinkLoops(t *testing.T) {
+	sketchPath, _ := paths.New("testdata", "SketchWithMultipleSymlinkLoops").Abs()
+
+	// The failure condition is New() never returning, testing for which requires setting up a timeout.
+	done := make(chan bool)
+	var sketch *Sketch
+	var err error
```

Comment:
```suggestion
	someSymlinkPath := sketchPath.Join("src", "some_symlink")
	// Create a recursive Sketch symlink
	err := os.Symlink(sketchPath.String(), someSymlinkPath.String())
	require.Nil(t, err, "This test must be run as administrator on Windows to have symlink creation privilege.")
	defer someSymlinkPath.Remove()

	someOtherSymlinkPath := sketchPath.Join("src", "some_other_symlink")
	os.Symlink(sketchPath.String(), someOtherSymlinkPath.String())
	defer someOtherSymlinkPath.Remove()

	// The failure condition is New() never returning, testing for which requires setting up a timeout.
	done := make(chan bool)
	var sketch *Sketch
```
Symlinks are kind of a pain on Windows. If you clone a repo that contains a symlink when not in adminstrator mode, what should be symlinks are just text files with the target path in them. In this case, it would mean the test fails due to `New()` not returning an error since there was no symlink loop.

For this reason, I have settled on the approach of generating symlink test data on the fly. That will fail if the test is run  on Windows while not in administrator mode, but in this case we can provide a helpful error message when the symlink creation errors, which should not occur otherwise.

If accepted, this should be accompanied by the removal of these files:
- `arduino/sketch/testdata/SketchWithMultipleSymlinkLoops/src/UpGoer1`
- `arduino/sketch/testdata/SketchWithMultipleSymlinkLoops/src/UpGoer2`

And the addition of this file:
`arduino/sketch/testdata/SketchWithMultipleSymlinkLoops/src/.gitkeep`
(alternatively, `arduino/sketch/testdata/SketchWithMultipleSymlinkLoops/src` could be generated on the fly)

Reply:
Ah, good to know! I'll fix both.